CF 1208F (#3131)
cpinitiative/usaco-guide: solutions/cf-1208F.mdx

---
id: cf-1208F
source: CF
title: Bits and Pieces
author: Dong Liu
---

## Explanation

> The answer is equivalent of enumerating $a_i$ and finding the optimal $x$ where $x=a_j\&a_k$ and $i<j<k$ and $a_i|x$ is maximum.

Reviewer: what's the "optimal" x?

Reviewer: maybe denote somewhere that | is the bitwise OR, thought it was divisibility for a second

Author: > what's the "optimal" x?

and $a_i|x$ is maximum.

Reviewer: what does that mean?

Note that $x$ can be binary searched since $2^b>2^0+2^1+...+2^{b-1}$, so we can loop through each $2^b$ in decreasing order and check if there exists an answer for $z$ such that $z$ is a superset of $x|2^b$. When we reach a bit $2^b$ such that it is contained in $a_i$, it doesn't matter whether it is in the final $x$ since $a_i|x$ already contains $2^b$.

> In other words, we have to find a pair $(j, k)$ such that $i<j<k$ and $a_j\&a_k$ is a superset of $x|2^b$. This is equivalent to finding $j$ and $k$ such that $j, k>i$, $a_j\supseteq x|2^b$, and $a_k\supseteq x|2^b$. This can be done by Sum Over Subsets DP (or supersets in this case) by maintaining $f_x$ and $g_x$, which stores the two largest $j$ such that $a_j\supseteq x$.

Reviewer: what the heck is the intuition for this

Author: Well, I guess the overall idea of the problem is to separate the terms in the formula (first term in this case). Observe that you can binary search (common? in maximizing OR problems). Then, you want j and k to be after $i$ and a_j, a_k has to be a superset of some value, so it would make sense to find the two greatest positions. Then, just optimize with SOS DP (because of the supersets part). The technique of maintaining the top two largest positions is pretty standard too I think.

## Implementation

**Time Complexity:** $\mathcal{O}(N\log A + A\log A)$ where $A=\max a_i$.

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

> const int N = 1e6, B = 21, A = 1 << B;

Reviewer: separate lines
also is B 21 or 20

Author: 21, since a_i <= 2e6

int main() {

> 	ios::sync_with_stdio(0), cin.tie(0);

Reviewer: remove if runs fine w/o

	int n; cin >> n;
	static int a[N], f[A], g[A];
	fill(f, f + A, -1), fill(g, g + A, -1);
	for (int i = 0; i < n; i++) {
		cin >> a[i];
		// second largest becomes largest, largest is updated to i
		f[a[i]] = g[a[i]], g[a[i]] = i;
	}

> 	for (int i = 0; i < B; i++)

Reviewer: braces

		for (int j = 0; j < A; j++)
			if (j & (1 << i)) {
				int k = j ^ (1 << i);
				// updating largest (g[k]) from superset k to subset j
				if (f[j] > g[k]) // case 1: on the right
					f[k] = g[k], g[k] = f[j];
				else if (f[j] > f[k]) // case 2: in the middle
					f[k] = f[j];
				// updating second largest (f[k]) from superset k to subset j
				if (g[j] > g[k])// case 1: on the right
					f[k] = g[k], g[k] = g[j];
				else if (g[j] > f[k])
					f[k] = g[j]; // case 2: in the middle
			}
	int ans = 0;
	for (int i = 0; i < n - 2; i++) {
		int c = 0;
		for (int j = B - 1; j >= 0; j--)
			// only consider if a[i] doesn't contain 2^j
			if ((a[i] & (1 << j)) == 0)
				// if the second largest is greater than i
				if (f[c | (1 << j)] > i)
					c |= 1 << j;
		ans = max(ans, a[i] | c);
	}

> 	cout << ans << '\n';

Reviewer: ```suggestion
	cout << ans << '\n';
```

Author: is there a difference?

Reviewer: oop

> 	return 0;

Reviewer: ```suggestion
```

---
id: cf-1208F
source: CF
title: Bits and Pieces
author: Dong Liu
---

## Explanation

Let $|$ denote the bitwise OR operation and $\&$ denote the bitwise AND operation.

> The answer is equivalent to enumerating $a_i$ and finding the optimal $x$ where $x=a_j\&a_k$ and $i<j<k$ and $a_i|x$ is maximum.

Reviewer: maybe smth like "equivalent to finding an $x$ such that [asdf] and $a_i|x$ is maximized"
also what does "enumerating" mean?

Author: > what does "enumerating" mean?

like looping over

Reviewer: actually i'd remove the enumerating part, doesn't seem to add anything

Author: how about replacing "enumerating" with "fixing an"?

Reviewer: why do we have to fix $i$ if it's going to go over all $i$ anyways
$i$, $j$, and $k$ can be basically any indices from what i'm seeing in this statement

Author: ohh I see

> Note that $x$ can be binary searched since $2^b>2^0+2^1+...+2^{b-1}$, so we can loop through each $2^b$ in decreasing order and check if there exists an answer for $z$ such that $z$ is a superset of $x|2^b$. When we reach a bit $2^b$ such that it is contained in $a_i$, it doesn't matter whether it is in the final $x$ since $a_i|x$ already contains $2^b$.

Reviewer 2: ```suggestion
Note that $x$ can be binary searched since $2^b>2^0+2^1+\dots+2^{b-1}$, so we can loop through each $2^b$ in decreasing order and check if there exists an answer for $z$ such that $z$ is a superset of $x|2^b$. When we reach a bit $2^b$ such that it is contained in $a_i$, it doesn't matter whether it is in the final $x$ since $a_i|x$ already contains $2^b$.
```

In other words, we have to find a pair $(j, k)$ such that $i<j<k$ and $a_j\&a_k$ is a superset of $x|2^b$. This is equivalent to finding $j$ and $k$ such that $j, k>i$, $a_j\supseteq x|2^b$, and $a_k\supseteq x|2^b$. This can be done by Sum Over Subsets DP (or supersets in this case) by maintaining $f_x$ and $g_x$, which stores the two largest $j$ such that $a_j\supseteq x$.

## Implementation

**Time Complexity:** $\mathcal{O}(N\log A + A\log A)$ where $A=\max a_i$.

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e6, B = 21, A = 1 << B;

int main() {
	int n; cin >> n;

> 	static int a[N], f[A], g[A];

Reviewer: you're using these locally already, why not just use vectors

	fill(f, f + A, -1), fill(g, g + A, -1);
	for (int i = 0; i < n; i++) {
		cin >> a[i];
		// second largest becomes largest, largest is updated to i

> 		f[a[i]] = g[a[i]], g[a[i]] = i;

Reviewer: separate lines

	}
	for (int i = 0; i < B; i++) {

> 		for (int j = 0; j < A; j++)

Reviewer: braces

			if (j & (1 << i)) {
				int k = j ^ (1 << i);
				// updating largest (g[k]) from superset k to subset j
				if (f[j] > g[k]) // case 1: on the right
					f[k] = g[k], g[k] = f[j];
				else if (f[j] > f[k]) // case 2: in the middle
					f[k] = f[j];
				// updating second largest (f[k]) from superset k to subset j
				if (g[j] > g[k]) // case 1: on the right
					f[k] = g[k], g[k] = g[j];
				else if (g[j] > f[k]) // case 2: in the middle
					f[k] = g[j];
			}
	}
	int ans = 0;
	for (int i = 0; i < n - 2; i++) {
		int c = 0;

> 		for (int j = B - 1; j >= 0; j--)

Reviewer: braces

---
id: cf-1208F
source: CF
title: Bits and Pieces
author: Dong Liu
---

## Explanation

Let $|$ denote the bitwise OR operation and $\&$ denote the bitwise AND operation.

> The answer is equivalent to finding an $x$ such that $x=a_j\&a_k$ and $i<j<k$ while $a_i|x$ is maximized.

Reviewer: ```suggestion
The answer is equivalent to finding an $x$ such that $x=a_j\&a_k$ and $a_i|x$ is maximized for some three indices $i<j<k$.
```

## Implementation

**Time Complexity:** $\mathcal{O}(N\log A + A\log A)$ where $A=\max a_i$.

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e6;
const int B = 21;
const int A = 1 << B;

int main() {
	int n;
	cin >> n;
	
	vector<int> a(n), f(A, -1), g(A, -1);
	for (int i = 0; i < n; i++) {
		cin >> a[i];
		// second largest becomes largest, largest is updated to i
		f[a[i]] = g[a[i]];
		g[a[i]] = i;
	}
	
	for (int i = 0; i < B; i++) {
		for (int j = 0; j < A; j++) {
			if (j & (1 << i)) {
				int k = j ^ (1 << i);
				// updating largest (g[k]) from superset k to subset j
				if (f[j] > g[k]) // case 1: on the right
					f[k] = g[k], g[k] = f[j];
				else if (f[j] > f[k]) // case 2: in the middle
					f[k] = f[j];
				// updating second largest (f[k]) from superset k to subset j
				if (g[j] > g[k]) // case 1: on the right
					f[k] = g[k], g[k] = g[j];

> 				else if (g[j] > f[k]) // case 2: in the middle

Reviewer: braces

	for (int i = 0; i < n - 2; i++) {
		int c = 0;
		for (int j = B - 1; j >= 0; j--) {

> 			// only consider if a[i] doesn't contain 2^j and if the second largest is greater than i

Reviewer: over 80 chars(?) if so, use a multiline comment
